Reimplement Resource.`_setup_local_to_scene` & deprecate signal

Mickeon opened the pull request:
Requires https://github.com/godotengine/godot/pull/81388.

This PR reimplements the virtual method `Resource._setup_local_to_scene`, that has been lost since #51970. It also gives it a much, **much** better description compared to 3.x:

Finally, deprecates the now redundant `setup_local_to_scene_requested` signal. 
This signal can only be useful when connecting it in the custom Resource's own `_init`. It _can_ be called manually with `setup_local_to_scene`. However, whether or not `setup_local_to_scene` should be exposed to the user in the first place is very up to debate to me (see my struggles attempting to update the description in https://github.com/godotengine/godot/pull/67072 and https://github.com/godotengine/godot/pull/67082.

##### I wish I could just remove it, to be honest. Only like, one or two people in the universe may be using it.

Mickeon committed:
Reimplement Resource.`_local_to_scene_setup` & remove workaround

Reimplements the virtual method _setup_local_to_scene, lost in #51970

Also deprecates the redundant `setup_local_to_scene_requested` signal.

Mickeon commented:
Rebased. Bumping this up.

KoBeWi commented:
The signal should be deprecated (and still functional) instead of being removed.

Mickeon commented:
So, wait, is the implementation on the PR not quite right? How to go about it?

Mickeon commented on core/io/resource.cpp:
```diff
@@ -379,8 +379,10 @@ Node *Resource::get_local_scene() const {
 }
 
 void Resource::setup_local_to_scene() {
-	// Can't use GDVIRTUAL in Resource, so this will have to be done with a signal
```
I... wow. I can't believe this. I was literally working on tweaking this PR on this exact moment.

YuriSizov replied:
I opened https://github.com/godotengine/godot/pull/81388, which I think should unblock you.

Mickeon commented:
This PR has been updated to to work with GDExtension, because the cyclic dependency has been thankfully solved.

It also is no longer compatibility-breaking, because the signal is deprecated, but functionality is still retained _(to my dismay)_.

YuriSizov commented:
> renames it to `_local_to_scene_setup` to better suit other similar virtual methods

Can you give an example? Typically a virtual method that needs to be implemented to change the behavior of a non-virtual methods is called exactly the same with the `_` prefix.

Mickeon commented:
> Can you give an example? Typically a virtual method that needs to be implemented to change the behavior of a non-virtual methods is called exactly the same with the `_` prefix.

I'm not sure why I made that decision last year. It's likely because `_local_to_scene_setup` sounded more natural. It's the method where the setup takes place, and... 

In hindsight I can't explain a good reason to keep it. Perhaps this is because virtual method names may have been standardized more across the board.
I could just bring back the old virtual method as it was without any conversion. I'm surprised no one mentioned it until the last second.

YuriSizov commented:
> I'm surprised no one mentioned it until the last second.

Well, we got to reviewing this PR only a few weeks back :P And it's more of a codestyle pass thing, while previous comments were about the implementation.

Mickeon commented:
Update the PR to change the name back to 3.x's equivalent.

Modified the description of `_setup_local_to_scene` slightly because I wasn't quite happy with it.
And added a deprecated message underneath `setup_local_to_scene` and `setup_local_to_scene_requested`.

KoBeWi commented on doc/classes/Resource.xml:
```diff
@@ -58,8 +73,8 @@
-				Emits the [signal setup_local_to_scene_requested] signal. If [member resource_local_to_scene] is set to [code]true[/code], this method is called from [method PackedScene.instantiate] by the newly duplicated resource within the scene instance.
-				For most resources, this method performs no logic of its own. Custom behavior can be defined by connecting [signal setup_local_to_scene_requested] from a script, [b]not[/b] by overriding this method.
+				Calls [method _setup_local_to_scene]. If [member resource_local_to_scene] is set to [code]true[/code], this method is automatically called from [method PackedScene.instantiate] by the newly duplicated resource within the scene instance.
+				[i]Deprecated.[/i] This method should only be called internally. Define your own methods or override [method _setup_local_to_scene], instead.
```
```suggestion
				[i]Deprecated.[/i] This method should only be called internally. Define your own methods or override [method _setup_local_to_scene] instead.
```

Mickeon replied:
I attempted to include the original argument of https://github.com/godotengine/godot/pull/67082. In a nutshell, there's no point using this method if you're going to define your own if necessary. It could probably be stripped away because it's... easy to infer.

KoBeWi replied:
The comma should still be removed.

KoBeWi commented on doc/classes/Resource.xml:
```diff
@@ -90,9 +105,10 @@
 			<description>
-				Emitted by the newly duplicated resource with [member resource_local_to_scene] set to [code]true[/code], when the scene is instantiated. Custom behavior can be defined by connecting this signal.
+				Emitted by a newly duplicated resource with [member resource_local_to_scene] set to [code]true[/code]. 
+				[i]Deprecated.[/i] This signal is only emitted when the resource is created. Override [method _setup_local_to_scene], instead.
```
```suggestion
				[i]Deprecated.[/i] This signal is only emitted when the resource is created. Override [method _setup_local_to_scene] instead.
```

KoBeWi commented on doc/classes/Resource.xml:
```diff
@@ -19,6 +19,21 @@
+		<method name="_setup_local_to_scene" qualifiers="virtual">
+			<return type="void" />
+			<description>
+				Override this method to customise the newly duplicated resource created from [method PackedScene.instantiate], if the original's [member resource_local_to_scene] is set to [code]true[/code].
```
```suggestion
				Override this method to customise the newly duplicated resource created from [method PackedScene.instantiate], if the original's [member resource_local_to_scene] is [code]true[/code].
```

AThousandShips replied:
```suggestion
				Override this method to customize the newly duplicated resource created from [method PackedScene.instantiate], if the original's [member resource_local_to_scene] is set to [code]true[/code].
```
US English should be used

KoBeWi commented on doc/classes/Resource.xml:
```diff
@@ -19,6 +19,21 @@
+			<return type="void" />
+			<description>
+				Override this method to customise the newly duplicated resource created from [method PackedScene.instantiate], if the original's [member resource_local_to_scene] is set to [code]true[/code].
+				[b]Example:[/b] Set a random [code]damage[/code] value to every local resource from an instantiated scene, excluding the original.
```
What "excluding the original" refers to? I did a test with this method in a tool script and every Resource instance has this called.

Mickeon replied:
"Excluding the original" refers to the resource every local resource has been duplicated from. The method is called on every new one, but the original Resource, basically the following in your scene, is not affected by this:

```Ini
[sub_resource type="Resource" id="Resource_gw2eg"]
resource_local_to_scene = true
script = ExtResource("1_khad7")
damage = 0
```

You may actually change this Resource at run time without affecting the duplicates. However, every **newly** instantiated scene will create its own local resources with the original properties in mind. 
This is worth knowing especially for Resources like Stylebox, or resources saved to disk.

Under the hood you may If you also feel like this is necessary information and/or have better ways to phrase it, feel free to suggest.

KoBeWi replied:
But if I open HolderOfLocal.tscn and save it, it will be modified. If I run the scene and print the value, it will also have a different value.
```GDScript
@export var res: Resource

func _ready() -> void:
	print(res.damage)
```
So when this method doesn't get called?

Mickeon replied:
That is... unexpected. Oh dear.
For now I'll remove the "except" stuff, but that feels like it should not happen. Maybe this bug, if it is one, only happens when a Resource is contained within a PackedScene?

Mickeon commented:
The final stretch. 

##### I'm leaving that conversation unsolved because it intrigues me.

AThousandShips commented on doc/classes/Resource.xml:
```diff
@@ -58,8 +73,8 @@
-				Emits the [signal setup_local_to_scene_requested] signal. If [member resource_local_to_scene] is set to [code]true[/code], this method is called from [method PackedScene.instantiate] by the newly duplicated resource within the scene instance.
-				For most resources, this method performs no logic of its own. Custom behavior can be defined by connecting [signal setup_local_to_scene_requested] from a script, [b]not[/b] by overriding this method.
+				Calls [method _setup_local_to_scene]. If [member resource_local_to_scene] is set to [code]true[/code], this method is automatically called from [method PackedScene.instantiate] by the newly duplicated resource within the scene instance.
+				[i]Deprecated.[/i] This method should only be called internally. Override [method _setup_local_to_scene], instead.
```
```suggestion
				[i]Deprecated.[/i] This method should only be called internally. Override [method _setup_local_to_scene] instead.
```
Reiterating this one

akien-mga commented:
Thanks!
